daemon/containerd/image.go

```diff
@@ -341,3 +339,131 @@ func (i *ImageService) getAllImagesWithRepository(ctx context.Context, ref refer
 	nameFilter := "^" + regexp.QuoteMeta(ref.Name()) + ":" + reference.TagRegexp.String() + "$"
 	return i.client.ImageService().List(ctx, "name~="+strconv.Quote(nameFilter))
 }
+
+func convertError(err error) error {
+	// TODO: Convert containerd error to Docker error
+	return err
+}
+
+// resolveAllReferences resolves the reference name or ID to an image and returns all the images with
+// the same target.
+func (i *ImageService) resolveAllReferences(ctx context.Context, refOrID string) (*containerdimages.Image, []containerdimages.Image, error) {
+	parsed, err := reference.ParseAnyReference(refOrID)
+	if err != nil {
+		return nil, nil, errdefs.InvalidParameter(err)
+	}
+	var dgst digest.Digest
+	var img *containerdimages.Image
+
+	if truncatedID.MatchString(refOrID) {
+		if d, ok := parsed.(reference.Digested); ok {
+			if cimg, err := i.images.Get(ctx, d.String()); err == nil {
+				img = &cimg
+				dgst = d.Digest()
+				if cimg.Target.Digest != dgst {
+					// Ambiguous image reference, use reference name
+					log.G(ctx).WithField("image", refOrID).WithField("target", cimg.Target.Digest).Warn("digest reference points to image with a different digest")
+					dgst = cimg.Target.Digest
+				}
+			} else if !cerrdefs.IsNotFound(err) {
+				return nil, nil, convertError(err)
+			} else {
+				dgst = d.Digest()
+			}
```

Comment:
(just a note when reviewing, wasn't sure which cases exactly would hit that path)

For the `^(sha256:)?([a-f0-9]{4,64})$` there are few distinct reference cases (`refOrID`):
1. `sha256:abcdef7890123456789012345678901234567890123456789012345678901234` (algo prefixed full digest)
2. `sha256:abce` (algo prefixed truncated digest)
3. `abcdef7890123456789012345678901234567890123456789012345678901234` (unprefixed full digest)
4. `abce` (unprefixed truncated digest)

This is what the reference is resolved to in each case (`reference.ParseAnyReference`):
1. `docker.io/library/sha256:abcdef7890123456789012345678901234567890123456789012345678901234`
  type: reference.taggedReference
2. `docker.io/library/sha256:abce`
  type: reference.taggedReference
3. `sha256:abcdef7890123456789012345678901234567890123456789012345678901234`
  type: reference.digestReference
4. `docker.io/library/abce`
  type: reference.repository


(https://go.dev/play/p/FU-Ysh-I2cI)

Reply:
1 Seems a bit surprising. I can add more unit tests around these cases to lessen the surprise.